AzureAD/microsoft-authentication-library-common-for-android
# Fixing potential deadlock state in executor service for interactive requests

### What
Fixing a potential deadlock issue in executor service for interactive request, by shutting down and recreating it.
This issue is reported in this ICM https://portal.microsofticm.com/imp/v3/incidents/details/264456886/home

### Why
The interactive request executor service can get into a state where an interactive request gets started in CommandDispatcher, but the actual AuthorizationActivity never starts. This can happen when user immediately press home button after starting an acquireToken (interactive) call to push the app in the background. Android 10 (API level 29) and higher place [restrictions ](https://developer.android.com/guide/components/activities/background-starts) on when apps can start activities when the app is running in the background. Because of this the AuthorizationActivity does not start and the request keep waiting for the Authorization Result. Even after broadcasting a "CANCEL_AUTHORIZATION_REQUEST" it doesnt finish the request as there are no listener for it.
This blocks any future interactive requests going through the broker as none of them can be assigned the thread. and the user sees an infinite spinner https://portal.microsofticm.com/imp/v3/incidents/details/264456886/home

### How
When we detect that there is no listener for "CANCEL_AUTHORIZATION_REQUEST" (which means that `AuthorizationActivity`/`AuthorizationFragment` hasn't started) but a previous request is still waiting for "RETURN_AUTHORIZATION_REQUEST_RESULT" (which means that an interactive request was started in the past), We identify the interactive thread getting stuck (deadlock). Once we run into this situation inorder to process incoming interactive requests, we shutdown the executor service and recreate it since it only has a single thread. This terminates the older thread and starts processing the interactive requests in new thread.

### Testing
**Repro** 
1. Launch MsalTestApp
2. Press AcquireToken
3. Quickly press home button
4. Go back to MsalTestApp to see if it gets stuck in infinite spinner case, if not try from step 2.
5. Press back button to cancel the request, 
6. Try acquireToken again
**Expected**: User should see the credentials page to perform interactive sign-in
**Actual**: User sees the spinner spinning. If you debug in CommandDispatcher.java, you can see that these requests are getting queued but there is already an active request which is still getting processed and doesnt finish.
**AfterFix**: The fix terminates the previous request and continues to process the new request and user sees the credential page to perform sign-in.
